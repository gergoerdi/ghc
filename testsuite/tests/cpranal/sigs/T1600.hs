module T1600 where


-- pretty strict
fac :: Int -> a ->  (a, Int)
fac n s | n < 2     = (s,1)
        | otherwise = case  fac (n-1) s of (s',n') -> let n'' = n*n' in n'' `seq` (s',n'')

-- lazier, but Int still has CPR
fac2 :: Int -> a ->  (a, Int)
fac2 n s | n < 2     = (s,1)
         | otherwise = case  fac (n-1) s of (s',n') -> (s',n'*n')

-- even lazier, no CPR on Int
fac3 :: Int -> a ->  (a, Int)
fac3 n s | n < 2     = (s,1)
         | otherwise = let (s',n') = fac (n-1) s in (s',n'*n')

facIO :: Int -> IO Int
facIO n | n < 2     = return 1
        | otherwise = do n' <- facIO (n-1); return (n*n')
