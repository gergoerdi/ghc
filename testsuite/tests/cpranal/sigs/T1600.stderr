
==================== Cpr signatures ====================
Lib.$trModule: #c1(#c1(#), #c1(#))
Lib.fac1: *c1(*, #c1(#))
Lib.fac2: *c1(*, #c1(#))
Lib.fac3: *c1(*, *c1(#))
Lib.facIO: *c1(*, #c1(#))


